### PR TITLE
[#27435] Describe the Source and Sink transform and urns in the protos.

### PR DESCRIPTION
**Pull request by lostluck**
Fill in a long standing gap in the proto documentation, by explicitly including the source and sink transforms among the StandardPTransform primitives.

I suspect a reason it went on so long without is that the Urns are largely in the Runner Proto, but everything about the transform is tied to the FnAPI implementation.

Since we no longer expect alternative FnAPI impelementations in Beam V2 that are not GRPC based, it's time to fill that gap.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @robertwb @byronellis 
cc: @kennknowles

**Comment by lostluck** [comment]
What I've put down is better than nothing, so I'm mostly concerned about inaccuracies based on my understanding from the Go SDK and implementing Prism.

The goal isn't to be pedantically precise, but convey the necessary high level view that will enable someone to actually implement the transforms in an SDK, and have it work.

File: model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/beam_runner_api.proto
Comment (kennknowles) [inline model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/beam_runner_api.proto:351]: I think the reason this isn't in the runner API proto is that it is a fn api detail, not a transform that has any meaning at the whole-graph level.

**Comment by kennknowles** [comment]
I think this touches on a philosophical difference. In my mind, the pipeline proto / runner API is for authoring parallel processing DAGs, while the Fn API is for invoking a UDF server. They have things in common but transforms should generally _not_ be punned between the two. Of course, transforms _have_ been punned between the two, leading to much confusion. In Dataflow v1 they were not punned, and it was good.

**Comment by lostluck** [comment]
I'm happy to put them in beam_fn_api.proto instead. I'm just tired of these being critical to function and being undiscoverable without looking at other SDKs.

It should be possible to construct a compliant Beam SDK without looking at how another SDK did it.

**Comment by kennknowles** [comment]
Agree. And yes put them in fn api.

**Review by kennknowles — APPROVED** [review]
No idea what "runner" has to do with anything but I'm not here to debate URNs and also that ship has sailed. Code matches URN and makes sense.

**Comment by lostluck** [comment]
> No idea what "runner" has to do with anything but I'm not here to debate URNs and also that ship has sailed. Code matches URN and makes sense.

Technically I think it represented how an SDK communicates with a Runner. But I agree with you, it's not a great name for the urn namespace, and it's too late to change these two. "beam:fnapi:runner:..." would have been a better prefix at least.  Something to fix in a Beam V3 years from now.

**Comment by lostluck** [comment]
I have an approval, and the example failure unrelated. I'll merge in  tomorrow morning if there are no objections / additional wordsmithing from other readers.

File: model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto
Comment (robertwb) [inline model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto:56]: I might say "It represents reading a PCollection from an external source (typically the Runner)" or "represents a stream of values coming in from an external source/over a data channel" as it's not the PCollection itself, but rather a descriptor of how to get the PCollection.

File: model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto
Comment (robertwb) [inline model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto:68]: We may allow other kinds of acceptable coders in the future, but you could say "must be a windowed value coder" generically.

File: model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto
Comment (robertwb) [inline model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto:74]: "remote port ProcessBundleDescriptor?"

Maybe "DataSink is a transform that sends PCollection elements to a remote port using the Data API."

**Review by robertwb — APPROVED** [review]
Thanks.

**Comment by kennknowles** [comment]
... yea and this is why keep namespaces short and sweet, like `beam:source:v1` or at most namespace it into `beam:transform:source:v1`. Code that looks legacy from day 1 makes me :-(

**Comment by kennknowles** [comment]
Kicked the tests.

File: model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto
Comment (lostluck) [reply model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto:68]: Done.

File: model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto
Comment (lostluck) [reply model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto:74]: Done.

File: model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto
Comment (lostluck) [reply model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto:56]: Done.

**Comment by lostluck** [comment]
Updated based on Robert's comments. Will merge once tests are passing again.

**Comment by lostluck** [comment]
Run Java PreCommit

**Comment by lostluck** [comment]
Run Java PreCommit

**Comment by lostluck** [comment]
Run Java PreCommit

**Pull request merged**
